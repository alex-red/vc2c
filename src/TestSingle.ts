import { convertFile } from ".";

const filePath = '../../tests/fixture/Input.ts'
const { file, result } = convertFile(filePath, __dirname, 'config/.compatible.vc2c.js');

console.log(result);