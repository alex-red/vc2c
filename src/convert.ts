import type ts from 'typescript'
import { getClassDeclarationNodes, hasComponentDecorator } from './utils'
import { convertASTResultToImport, runPlugins } from './plugins'
import { Vc2cOptions } from './options'
import { log } from './debug'
import { ASTResult } from './plugins/types'

const vueClassModules = [
  'vue-class-component',
  'vue-property-decorator'
]

export function convertAST (sourceFile: ts.SourceFile, options: Vc2cOptions, program: ts.Program): string {
  const tsModule = options.typescript

  log('check vue class library')
  const vueClassModuleImportStatement = sourceFile.statements
    .find((statement) => {
      if (tsModule.isImportDeclaration(statement)) {
        if (vueClassModules.includes((statement.moduleSpecifier as ts.StringLiteral).text)) {
          return true
        }
      }
      return false
    })
  if (!vueClassModuleImportStatement) {
    throw new Error('no vue class library in this file.')
  }

  log('check default export class')

  const otherStatements = sourceFile.statements
    .map((el) => el)
    .filter((el) =>
      !((tsModule.isClassDeclaration(el) && hasComponentDecorator(tsModule, el)) ||
      (tsModule.isImportDeclaration(el) && vueClassModules.includes((el.moduleSpecifier as ts.StringLiteral).text)) ||
      (tsModule.isImportDeclaration(el) && (el.moduleSpecifier as ts.StringLiteral).text === 'vue'))
    )

  let resultStatements = otherStatements;

  const classNodes = getClassDeclarationNodes(options.typescript, sourceFile)
  if (!classNodes) {
    throw new Error('no class components found')
  }

  let astResults: ASTResult<ts.Node>[] = [];

  for (let cNode of classNodes) {
    const result = runPlugins(cNode, options, program);
    astResults = astResults.concat(result.astResults);
    resultStatements.push(result.statement);
  }
  
  log('Make ImportDeclaration')
  resultStatements = resultStatements.concat(convertASTResultToImport(astResults, options));

  resultStatements = [
    ...resultStatements.filter((el) => tsModule.isImportDeclaration(el)),
    ...resultStatements.filter((el) => !tsModule.isImportDeclaration(el))
  ]

  log('output result code')
  const printer = tsModule.createPrinter()
  const result = printer.printFile(tsModule.factory.updateSourceFile(sourceFile, resultStatements))

  return result
}
