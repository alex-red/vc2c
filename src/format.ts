/* eslint-disable @typescript-eslint/no-var-requires */
import { Vc2cOptions } from './options'
import path from 'path'
import { log } from './debug'
import prettier from 'prettier/standalone'
// import prettierTypescriptParser from 'prettier/plugins/typescript'
import { existsFileSync } from './file'

export async function format (content: string, options: Vc2cOptions): Promise<string> {
  const isNode = typeof window === 'undefined'
  if (!isNode) {
    return await prettier.format(content, {
      plugins: ['typescript'],
      parser: 'typescript',
      semi: false,
      singleQuote: true
    })
  }

  const eslintConfigPath = path.resolve(options.root, options.eslintConfigFile)
  const prettierFormat = require('@timdp/prettier-eslint') as (config: unknown) => string
  const prettierEslintOpions = (existsFileSync(eslintConfigPath))
    ? {
      text: content,
      filePath: eslintConfigPath,
      prettierOptions: {
        parser: 'typescript'
      },
      fallbackPrettierOptions: {
        parser: 'typescript'
      }
    }
    : {
      text: content,
      filePath: undefined,
      eslintConfig: {
        parser: require.resolve('@typescript-eslint/parser'),
        parserOptions: {
          sourceType: 'module',
          ecmaFeatures: {
            jsx: false
          }
        },
        rules: {
          semi: ['error', 'never'],
          'padding-line-between-statements': [
            'error',
            { blankLine: 'always', prev: '*', next: 'export' },
            { blankLine: 'always', prev: 'const', next: '*' },
            { blankLine: 'always', prev: '*', next: 'const' }
          ]
        }
      },
      prettierOptions: {
        parser: 'typescript',
        Semicolons: false,
        singleQuote: true,
        trailingComma: 'none'
      },
      fallbackPrettierOptions: {
        parser: 'typescript',
        singleQuote: true,
        Semicolons: false,
        trailingComma: 'none'
      }
    }

  log('Format result code.....')

  return prettierFormat(prettierEslintOpions)
}
