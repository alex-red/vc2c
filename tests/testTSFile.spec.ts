import { convertFile } from '../src'
import { FileKind } from '../src/file'
import path from 'path'
import util from 'util'
import { exec } from 'child_process'

const execAsync = util.promisify(exec)

describe('testTSFile', () => {
  const defComponentPath = 'fixture/Input.ts';
  const multComponentPath = "fixture/MultipleComponents.ts";

  it('converts a file with a single default component', () => {
    const { file, result } = convertFile(defComponentPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(defComponentPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result).toMatchSnapshot()
  });

  it('converts a file with multiple components', () => {
    const { file, result } = convertFile(multComponentPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(multComponentPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result).toMatchSnapshot()
  });
})
