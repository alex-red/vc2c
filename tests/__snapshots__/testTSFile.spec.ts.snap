// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testTSFile converts a file with a component and non-component class 1`] = `
"import { defineComponent } from 'vue'

export class NonComponent {
  public testOne: string = '1'
  public testTwo: number = 2
}

/**
 * My basic tag
 */
export const BasicPropertyClass = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'BasicPropertyClass',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello }
  }
})
"
`;

exports[`testTSFile converts a file with a single default component 1`] = `
"import {
  computed,
  reactive,
  toRefs,
  ref,
  onBeforeUnmount,
  onUnmounted,
  watch,
  onMounted,
  defineComponent
} from 'vue'

/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: 'checked',
    event: 'change'
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'oao',
  setup(props, context) {
    /**
     * My count
     */
    const count = computed(() => {
      return store.state.count
    })

    const { a } = (() => {
      const a = 'pa'

      return toRefs(
        reactive({
          a: a
        })
      )
    })()

    const anotherComponent = ref<HTMLElement>(null)

    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.') // foo

    onBeforeUnmount(() => {
      context.emit('tearing down')
    })
    onUnmounted(() => {
      console.log('destroyed')
    })

    const click = (c: string) => {
      context.emit('click', c)
    }

    const value = computed({
      get: () => {
        return props.checked
      },
      set: (value: boolean) => {
        context.emit('change', value)
      }
    })

    watch(
      props.checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal)
        console.log(anotherComponent.value!)
      },
      { deep: true, immediate: true }
    )
    watch(msg.value, (val: string, newVal: string) => {
      console.log(val, newVal)
    })
    onMounted(() => {
      click('oao')
    })

    const removeItem = () => {
      if (props.foo > 10) {
        context.emit('remove-item', 10)
      } else {
        context.emit('remove-item', props.foo)
      }
    }

    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value)
    }

    removeItem()
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      msg,
      onBeforeUnmount,
      onUnmounted,
      click,
      value,
      checked,
      msg,
      onMounted,
      removeItem,
      hello
    }
  }
})
"
`;

exports[`testTSFile converts a file with multiple components 1`] = `
"import { defineComponent } from 'vue'

/**
 * My basic tag
 */
export const BasicPropertyClass = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'BasicPropertyClass',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello }
  }
})

export const AnotherComponent = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AnotherComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello2 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello2 }
  }
})

export const AThirdComponent = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AThirdComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello3 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello3 }
  }
})
"
`;

exports[`testTSFile converts a file with multiple components and one default 1`] = `
"import { defineComponent } from 'vue'

/**
 * My basic tag
 */
export const BasicPropertyClass = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'BasicPropertyClass',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello }
  }
})

export const AnotherComponent = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AnotherComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello2 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello2 }
  }
})

export default defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AThirdDefaultComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello3 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello3 }
  }
})
"
`;
